formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2017-2025 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("unused")
+@file:OptIn(ExperimentalUnsignedTypes::class)
+
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.*
+import kotlinx.serialization.cbor.internal.*
+
+/**
+ * Class representing single CBOR element.
+ * Can be [CborPrimitive], [CborMap] or [CborList].
+ *
+ * [CborElement.toString] properly prints CBOR tree as a human-readable representation.
+ * Whole hierarchy is serializable, but only when used with [Cbor] as [CborElement] is purely CBOR-specific structure
+ * which has a meaningful schemaless semantics only for CBOR.
+ *
+ * The whole hierarchy is [serializable][Serializable] only by [Cbor] format.
+ */
+@Serializable(with = CborElementSerializer::class)
+public sealed class CborElement(
+    /**
+     * CBOR tags associated with this element.
+     * Tags are optional semantic tagging of other major types (major type 6).
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    public val tags: ULongArray = ulongArrayOf()
```

Comment:
You probably want to keep out the default initializer to ensure that any concrete subclass passes the value (it is sealed so it is not a usability issue to require it to be passed along).

Reply:
Note that if tags are uncommon you should probably make it nullable to avoid the overhead of the empty array. Alternatively you could create an internal empty array value that is reused.

Reply:
i'd go for an internal empty array. It does not make sense to differentiate between a null value and an empty array